subprojects/gst-plugins-good/gst/rtp/gstrtph266pay.c

```diff
@@ -0,0 +1,754 @@
+  gobject_class->get_property = GST_DEBUG_FUNCPTR (_get_property);
+
+  g_object_class_install_property (gobject_class,
+      PROP_CONFIG_INTERVAL,
+      g_param_spec_int ("config-interval",
+          "Parameter Set send interval",
+          "Send VPS, SPS, PPS and APS at this interval (in seconds)"
+          "(0 = disabled, -1 = send with every IDR frame)",
+          -1, 3600, DEFAULT_CONFIG_INTERVAL,
+          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS)
+      );
+
+  gstelement_class->change_state = GST_DEBUG_FUNCPTR (_change_state);
+
+  gstrtpbasepayload_class->set_caps = GST_DEBUG_FUNCPTR (_set_caps);
+  gstrtpbasepayload_class->handle_buffer = GST_DEBUG_FUNCPTR (_handle_buffer);
+  gstrtpbasepayload_class->sink_event = GST_DEBUG_FUNCPTR (_sink_event);
+
+  gst_element_class_add_static_pad_template (gstelement_class, &src_template);
+  gst_element_class_add_static_pad_template (gstelement_class, &sink_template);
+
+  gst_element_class_set_static_metadata (gstelement_class, "RTP H266 payloader",
+      "Codec/Payloader/Network/RTP",
+      "Payload-encode H266 video into RTP packets (RFC 9328)",
+      "Carlos Falgueras García <cfalgueras@fluendo.com>");
+
+  GST_DEBUG_CATEGORY_INIT (GST_CAT_DEFAULT, "rtph266pay", 0,
+      "H266 RTP Payloader");
+}
+
+static void
+gst_rtp_h266_pay_init (GstRtpH266Pay * rtph266pay)
+{
+  rtph266pay->adapter = gst_adapter_new ();
+  g_queue_init (&rtph266pay->nalus);
+  rtph266pay->xps_id_nalu_map =
+      g_hash_table_new_full (g_direct_hash, g_direct_equal, NULL,
+      (GDestroyNotify) _nalu_free);
+  rtph266pay->alignment = ALIGNMENT_UNKOWN;
+  rtph266pay->last_xps_recv_ts = GST_CLOCK_TIME_NONE;
+  rtph266pay->last_xps_sent_ts = GST_CLOCK_TIME_NONE;
+  rtph266pay->config_interval = DEFAULT_CONFIG_INTERVAL;
+}
+
+static void
+_finalize (GObject * object)
+{
+  GstRtpH266Pay *rtph266pay = GST_RTP_H266_PAY (object);
+
+  g_clear_pointer (&rtph266pay->adapter, g_object_unref);
+  g_clear_pointer (&rtph266pay->xps_id_nalu_map, g_hash_table_destroy);
+  g_queue_clear_full (&rtph266pay->nalus, (GDestroyNotify) _nalu_free);
+
+  G_OBJECT_CLASS (parent_class)->finalize (object);
+}
+
+static void
+_set_property (GObject * object, guint prop_id,
+    const GValue * value, GParamSpec * pspec)
+{
+}
+
+static void
+_get_property (GObject * object, guint prop_id,
+    GValue * value, GParamSpec * pspec)
+{
+}
+
+static GstStateChangeReturn
+_change_state (GstElement * element, GstStateChange transition)
+{
+  GstRtpH266Pay *rtph266pay = GST_RTP_H266_PAY (element);
+  GstStateChangeReturn ret;
+
+  if (transition == GST_STATE_CHANGE_READY_TO_PAUSED)
+    gst_adapter_clear (rtph266pay->adapter);
+
+  ret = GST_ELEMENT_CLASS (parent_class)->change_state (element, transition);
+
+  // Chain up to the parent class first to avoid a race condition
+  // Check commit df724c410b02b82bd7db893d24e8572a06c2fcb1
+  if (transition == GST_STATE_CHANGE_PAUSED_TO_READY)
+    _clear_xps_cache (rtph266pay);
```

Comment:
Besides this is not what GStreamer does, why not clear the adapter here and nalu here, too?

Reply:
It's not necessary because we'll always come back to `GST_STATE_CHANGE_READY_TO_PAUSED` where we clear the NALU queue.